### PR TITLE
Migrates the java client authentication to Flight Auth v2

### LINKED ISSUE #3285: Java Client: Migrate off of deprecated newSession/refreshSessionToken to Flight Auth
DHE, as a client-side user of DHC, needs to use Flight Auth. The current version of `io.deephaven.client.impl.SessionImpl` still uses `SessionService::NewSession` and `SessionService::RefreshSessionToken` which are deprecated as they do not fit into either Flight Auth v1 or Flight Auth v2 molds.

There are tests in `FlightMessageRoundTripTest` that demonstrate how to use Flight Auth v1 and Flight Auth v2. However, they skirt around the need to worry about a session timing out and do not automatically refresh tokens (these are unit tests that aim to catch edge cases and are not concerned with true passage of time).

Fixing up `SessionImpl` to use the appropriate implementation for authentication will make client-behavior, under the reign of Flight Auth, more obvious, clear, and frankly, usable by DHE.

### PR DESCRIPTION
**Pull request by devinrsmith**
There are also plumbing improvements that allow constructing sessions with specific authentication values, and the sharing of authenticated channels without extraneous duplication of logic.

Fixes #3285

### CODE REVIEW

**Comment by devinrsmith** [comment]
This work was done in support of testing #3410. I wanted to take a bit further wrt DH URIs and BarrageTableResolver, but have time-bounded this work for now and filed https://github.com/deephaven/deephaven-core/issues/3421

File: java-client/barrage-dagger/src/main/java/io/deephaven/client/impl/BarrageSubcomponent.java [hunk 1]
```diff
@@ -33,6 +32,9 @@ interface Builder extends BarrageSessionFactoryBuilder {
 
         Builder allocator(@BindsInstance BufferAllocator bufferAllocator);
 
+        Builder authenticationTypeAndValue(
+                @BindsInstance @Nullable @Named("authenticationTypeAndValue") String authenticationTypeAndValue);
```
Comment (nbauernfeind) [inline java-client/barrage-dagger/src/main/java/io/deephaven/client/impl/BarrageSubcomponent.java:36]: Is this sufficient? Does this allow for bi-directional authentication schemes?

File: java-client/example-utilities/src/main/java/io/deephaven/client/examples/AuthenticationOptions.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Option(names = {"--psk"}, description = "The pre-shared key")
+    String psk;
+
+    @Option(names = {"--explicit"}, description = "The explicit authentication type and value")
+    String explicit;
+
+    public String toAuthenticationTypeAndValue() {
+        if (mtls != null && mtls) {
+            // todo
```
Comment (nbauernfeind) [inline java-client/example-utilities/src/main/java/io/deephaven/client/examples/AuthenticationOptions.java:19]: Is this a lost comment? Or should we add an issue tag?

File: java-client/flight/src/main/java/io/deephaven/client/impl/FlightSession.java [hunk 3]
```diff
@@ -20,6 +20,10 @@ public class FlightSession implements AutoCloseable {
 
     public static FlightSession of(SessionImpl session, BufferAllocator incomingAllocator,
             ManagedChannel channel) {
+        // Note: this pattern of FlightClient owning the ManagedChannel does not mesh well with the idea that some
```
Comment (nbauernfeind) [inline java-client/flight/src/main/java/io/deephaven/client/impl/FlightSession.java:23]: This comment is golden. Thanks; well said.

File: java-client/session/src/main/java/io/deephaven/client/impl/SessionImpl.java [hunk 4]
```diff
@@ -69,156 +63,77 @@
...
+        this.bearerChannel = Objects.requireNonNull(bearerChannel);
+        this.bearerHandler = Objects.requireNonNull(bearerHandler);
         this.exportTicketCreator = new ExportTicketCreator();
-        this.states = new ExportStates(this, sessionService, config.channel().table().withCallCredentials(credentials),
-                exportTicketCreator);
+        this.states = new ExportStates(this, bearerChannel.session(), bearerChannel.table(), exportTicketCreator);
         this.serialManager = TableHandleManagerSerial.of(this);
         this.batchManager = TableHandleManagerBatch.of(this, config.mixinStacktrace());
+        this.pingJob = config.executor().scheduleAtFixedRate(
+                () -> bearerChannel.config().getConfigurationConstants(
```
Comment (nbauernfeind) [inline java-client/session/src/main/java/io/deephaven/client/impl/SessionImpl.java:130]: Cristian pointed out that when the server was unavailable that the client retried in an excessive repetitive manner. Have you verified that this no longer happens? I believed it would require setting alternate grpc call options.

Also, Colin pointed out that possibly ConfigService might run on another host (say, for a cluster of instances) - should we add a ping grpc call directly to the `SessionService` in the event that we're hitting envoy and being routed to different hosts?

File: java-client/example-utilities/src/main/java/io/deephaven/client/examples/AuthenticationOptions.java (hunk 2 above)
Comment (devinrsmith) [reply java-client/example-utilities/src/main/java/io/deephaven/client/examples/AuthenticationOptions.java:19]: Meant to remove. We may eventually want proper string constants somewhere publicly accessible for authentication methods we support out of the box.

File: java-client/session/src/main/java/io/deephaven/client/impl/SessionImpl.java (hunk 4 above)
Comment (devinrsmith) [reply java-client/session/src/main/java/io/deephaven/client/impl/SessionImpl.java:130]: I've taken out that more "complicated" retry logic, and figured that using a fixed job that "pings" multiple times during the timeout interval is "good enough". (I don't see much value in some sort of backoff schedule... either we successfully reauth before the timeout interval, or we don't.)

I'd very much be in favor of a simple no-op "ping" RPC in `SessionService`; although, I think @niloc132 might have opinions? The general advice might be "use Handshake", but the code as written right now doesn't depend on Flight. Right now I'm essentially (ab)using `ConfigService#getConfigurationConstants` since I know the user is authorized for it.

File: java-client/barrage-dagger/src/main/java/io/deephaven/client/impl/BarrageSubcomponent.java (hunk 1 above)
Comment (devinrsmith) [reply java-client/barrage-dagger/src/main/java/io/deephaven/client/impl/BarrageSubcomponent.java:36]: No.

I think we want a working example we can interact with to come up with a better interface. Either way, the structure in main right now only supports anonymous authentication, and this adds on the ability for one-and-done type auth.

That said, I'm not generally happy w/ the structure of BarrageSubcomponent, FlightSubcomponent, SessionSubcomponent (I know I was the one who made them). We may want to remove some of these layers of abstractions or provide better interfaces for users to hook into.

**Review by nbauernfeind — APPROVED** [review]
